examples/speculative/speculative.cpp

```diff
@@ -94,9 +94,13 @@ int main(int argc, char ** argv) {
         }
     }
 
-    // tokenize the prompt
+
+    // Tokenize the prompt
+    const bool add_bos = llama_vocab_type(llama_get_model(ctx_tgt)) == LLAMA_VOCAB_TYPE_SPM;
```

Comment:
```suggestion
    const bool add_bos = llama_should_add_bos_token(llama_get_model(ctx_tgt));
```

Some SPM models don't want a BOS token. `llama_should_add_bos_token` takes metadata like `add_bos_token` into account. In the case where that metadata doesn't exist, it will default to `true` when SPM otherwise `false`.

We assume both models have the same `add_bos` setting, I'm not 100% sure that's correct. It's possible that it could be different, but also that would probably make speculation fail so maybe it should be an error when it doesn't match.

Reply:
The SPM check was copied from `main` before the new function was available; I've rebased on master and have replaced the SPM comparison with the correct function.

As for divergence between the two models with respect to adding BOS, would just checking if both have the same `add_bos` and if not error out like the vocab checks work?

Reply:
> As for divergence between the two models with respect to adding BOS, would just checking if both have the same 

Yes, I think that's very reasonable. The current iteration looks good to me, so I'm going to approve it. I didn't write the `speculative` example though so it's possible GG wants it to work differently. I'll let him merge it if he's happy with the current state (I'm pretty sure that will be the case, but just double checking).